apina-core/src/main/kotlin/fi/evident/apina/output/ts/AbstractTypeScriptGenerator.kt

```diff
@@ -217,23 +222,32 @@ abstract class AbstractTypeScriptGenerator(
         val signature = "${endpoint.name}Url($parameters): string"
 
         out.write(signature).write(" ").writeBlock {
-            out.write("return this.context.url(").writeValue(createConfig(endpoint, onlyUrl = true)).writeLine(");")
+            out.write("return this.context.url(").writeValue(createConfig(endpoint, onlyUrl = true, optionalTypeMode = settings.optionalTypeMode)).writeLine(");")
         }
     }
 
     private fun qualifiedTypeName(type: ApiType): String = when {
-        type is ApiType.Nullable -> qualifiedTypeName(type.type) + " | null"
-        type is ApiType.Primitive -> type.toTypeScript()
+        type is ApiType.Nullable -> when (settings.optionalTypeMode) {
+            OptionalTypeMode.UNDEFINED -> qualifiedTypeName(type.type) + " | undefined"
+            OptionalTypeMode.NULL -> qualifiedTypeName(type.type) + " | null"
+        }
+        type is ApiType.Primitive -> type.toTypeScript(settings.optionalTypeMode)
         type is ApiType.Array -> qualifiedTypeName(type.elementType) + "[]"
-        settings.isImported(ApiTypeName(type.toTypeScript())) -> type.toTypeScript()
-        else -> type.toTypeScript()
+        settings.isImported(ApiTypeName(type.toTypeScript(settings.optionalTypeMode))) -> type.toTypeScript(settings.optionalTypeMode)
+        else -> type.toTypeScript(settings.optionalTypeMode)
     }
 
     private fun discriminatedUnionMemberType(unionType: ApiTypeName, memberType: ClassDefinition) =
         "${unionType.name}_${memberType.type.name}"
 
     private fun parameterListCode(parameters: List<EndpointParameter>) =
-        parameters.joinToString(", ") { p -> p.name + ": " + qualifiedTypeName(p.type) }
+        parameters.sortedBy { p -> p.type is ApiType.Nullable }.joinToString(", ") { p ->
```

Comment:
I'm not sure but if sorting is done like this, is it possible that the parameters would end up in different order between generation runs?

Reply:
Kotlin's sorting is stable, so it is guaranteed to keep the relative order of elements for which the sort key is equal.

So that's not a problem, but I think the sorting of parameters into a different order by nullability is not only highly confusing, but also quite dangerous. Consider the following method on a controller:

```kotlin
@PostMapping
fun frobnicate(@RequestParam foo: Int, @RequestParam bar: Int) { ... }
```

and then an existing client calling that:

```typescript
endpoint.frobnicate(1, 2).substribe(...);
```

Now you make an innocent change on the server and make the first parameter nullable:

```kotlin
@PostMapping
fun frobnicate(@RequestParam foo: Int?, @RequestParam bar: Int) { ... }
```

Now the client code will still compile, but it has different semantics! This is not a happy place to be.

I think the only sane way to compile this is as I wrote in my previous comment, perhaps too unclearly. Trailing nullable parameters can be made completely nullable, others can't. That's more or less the same that most programming languages do anyway: you can have default values for parameters at the end, but not in the middle.

Furthermore, as I mentioned, this part of the code does not need to depend on `optionalTypeMode`. For method parameters we can always translate optionals as allowing both `null` and `undefined` without downsides.

Reply:
So I was thinking about something like this. Written here in browser and completely untested, so probably has some silly bugs, but I guess the idea is clear: we ind out the index after which optional parameters are allowed and then make parameters after that optional.

```kotlin
private fun parameterListCode(parameters: List<EndpointParameter>): String {
    val firstOptionalIndex = parameters.indexOfLast { it.type !is ApiType.Nullable } + 1
    
    return parameters.withIndex().joinToString(", ") { (index, p) ->
        if (p.type is ApiType.Nullable) {
            val type = qualifiedTypeName(p.type.unwrapNullable())
            if (index >= firstOptionalIndex)
                "${p.name}?: $type | null"
            else 
                "${p.name}: $type | null | undefined"
        } else {
            "${p.name}: ${qualifiedTypeName(p.type)}"
        }
    }
}
```

Reply:
Yeah this seems quite dangerous. But actually we cannot use optional props, because:
`foo(param?: string | null, otherParam: string)` would still complain about required parameter following the optional. I'll remove this part and let's just use `prop: string | undefined` for now?

Reply:
Yeah, I mentioned both times that only _trailing_ nullable parameters can be made optional. :smile:

Reply:
Yeah... somehow I missed that.. 😄  But your suggestion seems to work great! 👍